--- /home/timothy/miniforge3/envs/VEBA-binning-prokaryotic_env_v2/lib/python3.8/site-packages/mdmcleaner/review_refdbcontams.py	2024-11-26 08:08:51.341792033 -0500
+++ review_refdbcontams.py	2023-01-12 05:33:49.000000000 -0500
@@ -302,7 +302,7 @@
 			blastrecords = [self.blastxjobs[x].seqrecord[0] for x in self.blastxjobs if len(self.blastxjobs[x].seqrecord[0]) < 100000] #for now skipping reference contigs larger than 100 kb (takes too long to blastx). TODO: in such cases, search for ribosomal & other markergenes to verify classification!
 			sys.stderr.write("\nblasting {} entries with blastx against reference proteins (another {} entries were too long to blastx efficiently\n".format(len(blastrecords), len(self.blastxjobs) - len(blastrecords)))
 			if len(blastrecords) == 0:
+				return [None]
-				return
 			basic_blastarglist = [ (blastrecords, blastdb, "diamond blastx") for blastdb in self.blastxdbs ]
 			# ~ import pdb; pdb.set_trace()
 			import time
--- /home/timothy/miniforge3/envs/VEBA-binning-prokaryotic_env_v2/lib/python3.8/site-packages/mdmcleaner/getdb.py	2024-11-21 10:01:47.409683731 -0500
+++ getdb.py	2023-01-12 05:33:49.000000000 -0500
@@ -47,9 +47,9 @@
 #todo: find common names for gtdb and ncbi dbs to simplify things
 #todo: find actual names for ncbi dbs
 dbfiles = { "gtdb" : {	"protblastdbs" : ["gtdbplus_protdb.dmnd"], \
+						"nucblastdbs" : ["concat_refgenomes", "SILVA_138.2_SSURef_NR99_tax_silva", "SILVA_138.2_LSURef_NR99_tax_silva"] ,\
+						"ssu_nucblastdbs" : ["concat_refgenomes", "SILVA_138.2_SSURef_NR99_tax_silva"], \
+						"lsu_nucblastdbs" : ["concat_refgenomes", "SILVA_138.2_LSURef_NR99_tax_silva"], \
-						"nucblastdbs" : ["concat_refgenomes", "SILVA_138.1_SSURef_NR99_tax_silva", "SILVA_138.1_LSURef_NR99_tax_silva"] ,\
-						"ssu_nucblastdbs" : ["concat_refgenomes", "SILVA_138.1_SSURef_NR99_tax_silva"], \
-						"lsu_nucblastdbs" : ["concat_refgenomes", "SILVA_138.1_LSURef_NR99_tax_silva"], \
 						"genome_nucblastdbs" : ["concat_refgenomes"], \
 						"mdmdbs" : ["gtdb_all.accession2taxid.sorted", "gtdb_taxonomy_br.json.gz", "gtdb_lcawalkdb_br.db"] },\
 			"ncbi" : {  "protblastdbs" : ["nr"], \
